tests/api/test_fqn.py

```diff
@@ -0,0 +1,141 @@
+from unittest import mock
+from unittest.mock import MagicMock
+
+import pytest
+from snowflake.cli.api.exceptions import FQNInconsistencyError, FQNNameError
+from snowflake.cli.api.identifiers import FQN
+from snowflake.cli.api.project.schemas.streamlit.streamlit import Streamlit
+
+
+def test_attributes():
+    fqn = FQN(name="object_name", database="database_name", schema="schema_name")
+    assert fqn.database == "database_name"
+    assert fqn.schema == "schema_name"
+    assert fqn.name == "object_name"
+
+
+def test_identifier():
+    fqn = FQN(name="object_name", database="database_name", schema="schema_name")
+    assert fqn.identifier == "database_name.schema_name.object_name"
+
+
+def test_quoted_identifier():
+    fqn = FQN(name='"my object"', database="database_name", schema="schema_name")
+    assert fqn.identifier == 'database_name.schema_name."my object"'
+
+
+@pytest.mark.parametrize(
+    "fqn, expected",
+    [
+        (
+            FQN(name="object_name", database="database_name", schema="schema_name"),
+            "DATABASE_NAME.SCHEMA_NAME.OBJECT_NAME",
+        ),
+        (
+            FQN(name='"quoted name"', database="database_name", schema="schema_name"),
+            "DATABASE_NAME.SCHEMA_NAME.quoted name",
+        ),
+    ],
+)
+def test_url_identifier(fqn, expected):
+    assert fqn.url_identifier == expected
+
+
+def test_set_database():
+    fqn = FQN(name="object_name", database="database_name", schema="schema_name")
+    fqn.set_database("foo")
+    assert fqn.database == "foo"
+
+
+def test_set_schema():
+    fqn = FQN(name="object_name", database="database_name", schema="schema_name")
+    fqn.set_schema("foo")
+    assert fqn.schema == "foo"
+
+
+@pytest.mark.parametrize(
```

Comment:
I think you missed testing the nested double quote case, e.g. `"abc""this is in nested quotes"""` (doubling the inside quotes is Snowflake's escaping mechanism)